File: dsa/medium/13. Min Height BST/min_height_bst.go
package main

// start the root for the tree from the middle of the vector
func MinHeightBST(array []int) *BST {
	if len(array) == 0 {
		return nil
	}
	mid := len(array) / 2
	tree := &BST{Value: array[mid]}

	arrayLeft := array[:mid]
	arrayRight := array[mid+1:]

	tree.Left = MinHeightBST(arrayLeft)
	tree.Right = MinHeightBST(arrayRight)

	return tree
}

type BST struct {
	Value int

	Left  *BST
	Right *BST
}

func (tree *BST) Insert(value int) *BST {
	if value < tree.Value {
		if tree.Left == nil {
			tree.Left = &BST{Value: value}
		} else {
			tree.Left.Insert(value)
		}
	} else {
		if tree.Right == nil {
			tree.Right = &BST{Value: value}
		} else {
			tree.Right.Insert(value)
		}
	}
	return tree
}

File: dsa/medium/13. Min Height BST/min_height_bst.cpp
#include <vector>

class BST {
public:
  int value;
  BST *left;
  BST *right;

  BST(int value) {
    this->value = value;
    left = nullptr;
    right = nullptr;
  }

  void insert(int value) {
    if (value < this->value) {
      if (left == nullptr) {
        left = new BST(value);
      } else {
        left->insert(value);
      }
    } else {
      if (right == nullptr) {
        right = new BST(value);
      } else {
        right->insert(value);
      }
    }
  }
};

// start the root for the tree from the middle of the vector
BST *minHeightBst(std::vector<int> array) {
  if (array.size() == 0) {
    return nullptr;
  }

  int mid = array.size() / 2;
  BST *root = new BST(array[mid]);

  std::vector<int> left(array.begin(), array.begin() + mid);
  std::vector<int> right(array.begin() + mid + 1, array.end());

  root->left = minHeightBst(left);
  root->right = minHeightBst(right);

  return root;
}
